crates/ruff_server/src/server/main_loop.rs

```diff
@@ -0,0 +1,217 @@
+            .workspace
+            .as_ref()
+            .and_then(|workspace| workspace.did_change_watched_files)
+            .and_then(|watched_files| watched_files.dynamic_registration)
+            .unwrap_or_default();
+        if dynamic_registration {
+            // Register all dynamic capabilities here
+
+            // `workspace/didChangeWatchedFiles`
+            // (this registers the configuration file watcher)
+            let params = lsp_types::RegistrationParams {
+                registrations: vec![lsp_types::Registration {
+                    id: "ruff-server-watch".into(),
+                    method: "workspace/didChangeWatchedFiles".into(),
+                    register_options: Some(
+                        serde_json::to_value(DidChangeWatchedFilesRegistrationOptions {
+                            watchers: vec![
+                                FileSystemWatcher {
+                                    glob_pattern: types::GlobPattern::String(
+                                        "**/.ruff.toml".into(),
+                                    ),
+                                    kind: None,
+                                },
+                                FileSystemWatcher {
+                                    glob_pattern: types::GlobPattern::String("**/ruff.toml".into()),
+                                    kind: None,
+                                },
+                                FileSystemWatcher {
+                                    glob_pattern: types::GlobPattern::String(
+                                        "**/pyproject.toml".into(),
+                                    ),
+                                    kind: None,
+                                },
+                            ],
+                        })
+                        .unwrap(),
+                    ),
+                }],
+            };
+
+            let response_handler = |_: &Client, ()| {
+                tracing::info!("Configuration file watcher successfully registered");
+            };
+
+            if let Err(err) = client.send_request::<lsp_types::request::RegisterCapability>(
+                &self.session,
+                params,
+                response_handler,
+            ) {
+                tracing::error!(
+                    "An error occurred when trying to register the configuration file watcher: {err}"
+                );
+            }
+        } else {
+            tracing::warn!(
+                "LSP client does not support dynamic capability registration - automatic configuration reloading will not be available."
+            );
+        }
+    }
+}
+
+/// An action that should be performed on the main loop.
+#[derive(Debug)]
+pub enum Action {
+    /// Send a response to the client
+    SendResponse(lsp_server::Response),
+}
+
+#[derive(Debug)]
+pub enum Event {
+    /// An incoming message from the LSP client.
+    Message(lsp_server::Message),
+
+    Action(Action),
+}
```

Comment:
If we can't re-use the client implementation, we should probably inline the single enum variant in `Event`:
```rs
#[derive(Debug)]
pub enum Event {
    /// An incoming message from the LSP client.
    Message(lsp_server::Message),

    /// Send a response to the client
    SendResponse(lsp_server::Response),
}
```